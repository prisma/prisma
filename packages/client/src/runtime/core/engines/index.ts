export { BinaryEngine } from './binary/BinaryEngine'
export {
  type BatchTransactionOptions,
  type Engine,
  type EngineConfig,
  type GraphQLQuery,
  type InteractiveTransactionOptions,
  type TransactionOptions,
} from './common/Engine'
export * from './common/types/EngineValidationError'
export type { LogEmitter } from './common/types/Events'
export * from './common/types/JsonProtocol'
export type { Metric, MetricHistogram, MetricHistogramBucket, Metrics } from './common/types/Metrics'
export type { IsolationLevel, Options, TransactionHeaders } from './common/types/Transaction'
export { DataProxyEngine } from './data-proxy/DataProxyEngine'
export type { Fetch } from './data-proxy/utils/request'
export { LibraryEngine } from './library/LibraryEngine'
export * as NodeAPILibraryTypes from './library/types/Library'
