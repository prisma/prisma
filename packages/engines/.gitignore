node_modules
query-engine*
libquery_engine*
introspection-engine*
migration-engine*
prisma-fmt*