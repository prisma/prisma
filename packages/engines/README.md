# `@prisma/engines`

⚠️ **Warning**: This package is intended for Prisma's internal use.
Its release cycle does not follow SemVer, which means we might release breaking changes (change APIs, remove functionality) without any prior warning.

The postinstall hook of this package downloads all Prisma engines available for the current platform, namely the Query Engine and the Migration Engine from the Prisma CDN.

The engines version to be downloaded are directly determined by the version of its `@prisma/engines-version` dependency.

You should probably not use this package directly, but instead use one of these:

- [`prisma` CLI](https://www.npmjs.com/package/prisma)
- [`@prisma/client`](https://www.npmjs.com/package/@prisma/client)
