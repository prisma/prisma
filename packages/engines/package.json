{
  "name": "@prisma/engines",
  "version": "0.0.0",
  "description": "This package is intended for Prisma's internal use",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "license": "Apache-2.0",
  "author": "Tim Suchanek <suchanek@prisma.io>",
  "devDependencies": {
    "@prisma/debug": "workspace:*",
    "@prisma/engines-version": "4.9.0-39.df61b5f5ab3f27359e273f1cba4e5a9c7cb8485a",
    "@prisma/fetch-engine": "workspace:*",
    "@prisma/get-platform": "workspace:*",
    "@swc/core": "1.3.14",
    "@swc/jest": "0.2.24",
    "@types/jest": "29.2.4",
    "@types/node": "16.18.11",
    "execa": "5.1.1",
    "jest": "29.3.1",
    "typescript": "4.8.4"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/prisma/prisma.git",
    "directory": "packages/engines"
  },
  "scripts": {
    "dev": "DEV=true node -r esbuild-register helpers/build.ts",
    "build": "node -r esbuild-register helpers/build.ts",
    "test": "jest",
    "prepublishOnly": "pnpm run build",
    "postinstall": "node scripts/postinstall.js"
  },
  "files": [
    "dist",
    "download",
    "scripts"
  ],
  "sideEffects": false
}
