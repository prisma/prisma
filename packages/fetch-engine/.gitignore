node_modules
dist
build
runtime
download
cli
.DS_Store
prisma
schema-inferrer-bin
test
sandbox
pnpm-lock.yaml
