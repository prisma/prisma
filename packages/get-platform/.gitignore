node_modules
.DS_Store
query-engine*
yarn-error.log
pnpm-debug.log

*.db

dist/

copy-photon.sh
*.tsbuildinfo