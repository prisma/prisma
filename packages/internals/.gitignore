node_modules
query-engine*
libquery_engine*
introspection-engine*
migration-engine*
prisma-fmt*
libquery*
/engines
/debug.d.ts
/debug.js
/debug.js.map
pnpm-lock.yaml
yarn-error.log
.DS_Store

