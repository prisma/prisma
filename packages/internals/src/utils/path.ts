import path from 'path'

/**
 * Normalize `filePath` to use forward slashes as a separator. `filePath` is
 * treated as a path specific to the current platform, so backslashes will only
 * be replaced with forward slashes on Windows. On other platforms, where a
 * backslash is a valid filename character, it will be treated as such and will
 * not be replaced.
 */
export function pathToPosix(filePath: string): string {
  if (path.sep === path.posix.sep) {
    return filePath
  }
  return filePath.split(path.sep).join(path.posix.sep)
}

/**
 * Returns the longest common path ancestor of the two paths (which may also be equal to one or both of them).
 * If they don't share a common ancestor (which may be the case on Windows if both paths are on different disks),
 * returns `undefined`.
 */
export function longestCommonPathPrefix(pathA: string, pathB: string): string | undefined {
  if (!path.isAbsolute(pathA) || !path.isAbsolute(pathB)) {
    throw new Error('longestCommonPathPrefix expects absolute paths')
  }

  if (process.platform === 'win32' && (pathA.startsWith('\\\\') || pathB.startsWith('\\\\'))) {
    // Make both paths namespaced if at least one of them is.
    pathA = path.toNamespacedPath(pathA)
    pathB = path.toNamespacedPath(pathB)
  }

  const commonPrefix = longestCommonPrefix(pathA.split(path.sep), pathB.split(path.sep)).join(path.sep)

  if (commonPrefix === '') {
    return process.platform === 'win32' ? undefined : '/'
  }

  if (process.platform === 'win32' && ['\\', '\\\\?', '\\\\.'].includes(commonPrefix)) {
    return undefined
  }

  if (process.platform === 'win32' && commonPrefix.endsWith(':')) {
    // Disk specifier without a backslash at the end is not an absolute path on windows,
    // it refers to current working directory on that disk.
    return commonPrefix + '\\'
  }

  return commonPrefix
}

function longestCommonPrefix<T>(sequenceA: T[], sequenceB: T[]): T[] {
  const maxLen = Math.min(sequenceA.length, sequenceB.length)
  let sliceLen = 0

  while (sliceLen <= maxLen && sequenceA[sliceLen] === sequenceB[sliceLen]) {
    sliceLen++
  }

  return sequenceA.slice(0, sliceLen)
}
