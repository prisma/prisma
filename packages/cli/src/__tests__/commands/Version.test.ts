import { enginesVersion, getCliQueryEngineBinaryType } from '@prisma/engines'
import { BinaryType, download } from '@prisma/fetch-engine'
import { getPlatform } from '@prisma/get-platform'
import { engineEnvVarMap, jestConsoleContext, jestContext } from '@prisma/internals'
import { ensureDir } from 'fs-extra'
import path from 'path'

import packageJson from '../../../package.json'

const ctx = jestContext.new().add(jestConsoleContext()).assemble()
const testIf = (condition: boolean) => (condition ? test : test.skip)
const useNodeAPI = getCliQueryEngineBinaryType() === BinaryType.libqueryEngine
const version = '39190b250ebc338586e25e6da45e5e783bc8a635'

describe('version', () => {
  // Node-API Tests

  testIf(useNodeAPI)('basic version (Node-API)', async () => {
    const data = await ctx.cli('--version')
    expect(cleanSnapshot(data.stdout)).toMatchSnapshot()
  })

  testIf(useNodeAPI)(
    'version with custom binaries (Node-API)',
    async () => {
      const enginesDir = path.join(__dirname, 'version-test-engines')
      await ensureDir(enginesDir)
      const binaryPaths = await download({
        binaries: {
          'introspection-engine': enginesDir,
          'migration-engine': enginesDir,
          'prisma-fmt': enginesDir,
          'libquery-engine': enginesDir,
        },
        version,
        failSilent: false,
      })
      // This Omits query-engine from the map
      const { ['query-engine']: qe, ...envVarMap } = engineEnvVarMap

      const platform = await getPlatform()

      for (const engine in envVarMap) {
        const envVar = envVarMap[engine]
        process.env[envVar] = binaryPaths[engine][platform]
        // console.debug(`Setting ${envVar} to ${binaryPaths[engine][platform]}`)
      }

      const data = await ctx.cli('--version')
      expect(cleanSnapshot(data.stdout, version)).toMatchSnapshot()

      // cleanup
      for (const engine in envVarMap) {
        const envVar = envVarMap[engine]
        delete process[envVar]
      }
    },
    50_000,
  )

  // Binary Tests

  testIf(!useNodeAPI)(
    'basic version',
    async () => {
      const data = await ctx.cli('--version')
      expect(cleanSnapshot(data.stdout)).toMatchSnapshot()
    },
    10_000,
  )

  testIf(!useNodeAPI)(
    'version with custom binaries',
    async () => {
      const enginesDir = path.join(__dirname, 'version-test-engines')
      await ensureDir(enginesDir)
      const binaryPaths = await download({
        binaries: {
          'introspection-engine': enginesDir,
          'migration-engine': enginesDir,
          'prisma-fmt': enginesDir,
          'query-engine': enginesDir,
        },
        version,
        failSilent: false,
      })

      const platform = await getPlatform()
      const { ['libquery-engine']: qe, ...envVarMap } = engineEnvVarMap
      for (const engine in envVarMap) {
        const envVar = envVarMap[engine]
        process.env[envVar] = binaryPaths[engine][platform]
        // console.debug(`Setting ${envVar} to ${binaryPaths[engine][platform]}`)
      }

      const data = await ctx.cli('--version')
      expect(cleanSnapshot(data.stdout, version)).toMatchSnapshot()

      // cleanup
      for (const engine in envVarMap) {
        const envVar = envVarMap[engine]
        delete process[envVar]
      }
    },
    50_000,
  )
})

function cleanSnapshot(str: string, versionOverride?: string): string {
  // sanitize engine path
  // Query Engine (Node-API) : libquery-engine e996df5d66a2314d1da15d31047f9777fc2fbdd9 (at ../../home/runner/work/prisma/prisma/node_modules/.pnpm/@prisma+engines@3.11.0-41.e996df5d66a2314d1da15d31047f9777fc2fbdd9/node_modules/@prisma/engines/libquery_engine-TEST_PLATFORM.LIBRARY_TYPE.node)
  // +                                                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  // Query Engine (Node-API) : libquery-engine 5a2e5869b69a983e279380ec68596b71beae9eff (at ../../cli/src/__tests__/commands/version-test-engines/libquery_engine-TEST_PLATFORM.LIBRARY_TYPE.node, resolved by PRISMA_QUERY_ENGINE_LIBRARY)
  // =>                                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  // Query Engine (Node-API) : libquery-engine e996df5d66a2314d1da15d31047f9777fc2fbdd9 (at sanitized_path/libquery_engine-TEST_PLATFORM.LIBRARY_TYPE.node)
  //                                                                                    ^^^^^^^^^^^^^^^^^^^
  str = str.replace(/\(at (.*engines)(\/|\\)/g, '(at sanitized_path/')

  // TODO: replace '[a-z0-9]{40}' with 'ENGINE_VERSION'.
  // Currently, the engine version of @prisma/prisma-fmt-wasm isn't necessarily the same as the enginesVersion
  str = str.replace(/([0-9]+\.[0-9]+\.[0-9]+-[0-9]+\.)([a-z0-9-]+)/g, 'CLI_VERSION.ENGINE_VERSION')

  // replace engine version hash
  const defaultEngineVersion = enginesVersion
  const currentEngineVersion = versionOverride ?? enginesVersion
  str = str.replace(new RegExp(currentEngineVersion, 'g'), 'ENGINE_VERSION')
  str = str.replace(new RegExp(defaultEngineVersion, 'g'), 'ENGINE_VERSION')
  str = str.replace(new RegExp('workspace:\\*', 'g'), 'ENGINE_VERSION')

  // replace studio version
  str = str.replace(packageJson.devDependencies['@prisma/studio-server'], 'STUDIO_VERSION')

  // sanitize windows specific engine names
  str = str.replace(/\.exe/g, '')

  return str
}
