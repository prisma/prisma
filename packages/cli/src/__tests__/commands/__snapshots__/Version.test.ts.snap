// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`version basic version (Node-API) 1`] = `
prisma                  : 0.0.0
@prisma/client          : Not found
Current platform        : TEST_PLATFORM
Query Engine (Node-API) : libquery-engine ENGINE_VERSION (at sanitized_path/libquery_engine-TEST_PLATFORM.LIBRARY_TYPE.node)
Migration Engine        : migration-engine-cli ENGINE_VERSION (at sanitized_path/migration-engine-TEST_PLATFORM)
Introspection Engine    : introspection-core ENGINE_VERSION (at sanitized_path/introspection-engine-TEST_PLATFORM)
Format Binary           : prisma-fmt ENGINE_VERSION (at sanitized_path/prisma-fmt-TEST_PLATFORM)
Format Wasm             : @prisma/prisma-fmt-wasm CLI_VERSION.ENGINE_VERSION
Default Engines Hash    : ENGINE_VERSION
Studio                  : STUDIO_VERSION
`;

exports[`version version with custom binaries (Node-API) 1`] = `
prisma                  : 0.0.0
@prisma/client          : Not found
Current platform        : TEST_PLATFORM
Query Engine (Node-API) : libquery-engine ENGINE_VERSION (at sanitized_path/libquery_engine-TEST_PLATFORM.LIBRARY_TYPE.node, resolved by PRISMA_QUERY_ENGINE_LIBRARY)
Migration Engine        : migration-engine-cli ENGINE_VERSION (at sanitized_path/migration-engine-TEST_PLATFORM, resolved by PRISMA_MIGRATION_ENGINE_BINARY)
Introspection Engine    : introspection-core ENGINE_VERSION (at sanitized_path/introspection-engine-TEST_PLATFORM, resolved by PRISMA_INTROSPECTION_ENGINE_BINARY)
Format Binary           : prisma-fmt ENGINE_VERSION (at sanitized_path/prisma-fmt-TEST_PLATFORM, resolved by PRISMA_FMT_BINARY)
Format Wasm             : @prisma/prisma-fmt-wasm CLI_VERSION.ENGINE_VERSION
Default Engines Hash    : ENGINE_VERSION
Studio                  : STUDIO_VERSION
`;

exports[`version basic version 1`] = `
prisma                : 0.0.0
@prisma/client        : Not found
Current platform      : TEST_PLATFORM
Query Engine (Binary) : query-engine ENGINE_VERSION (at sanitized_path/query-engine-TEST_PLATFORM)
Migration Engine      : migration-engine-cli ENGINE_VERSION (at sanitized_path/migration-engine-TEST_PLATFORM)
Introspection Engine  : introspection-core ENGINE_VERSION (at sanitized_path/introspection-engine-TEST_PLATFORM)
Format Binary         : prisma-fmt ENGINE_VERSION (at sanitized_path/prisma-fmt-TEST_PLATFORM)
Format Wasm           : @prisma/prisma-fmt-wasm CLI_VERSION.ENGINE_VERSION
Default Engines Hash  : ENGINE_VERSION
Studio                : STUDIO_VERSION
`;

exports[`version version with custom binaries 1`] = `
prisma                : 0.0.0
@prisma/client        : Not found
Current platform      : TEST_PLATFORM
Query Engine (Binary) : query-engine ENGINE_VERSION (at sanitized_path/query-engine-TEST_PLATFORM, resolved by PRISMA_QUERY_ENGINE_BINARY)
Migration Engine      : migration-engine-cli ENGINE_VERSION (at sanitized_path/migration-engine-TEST_PLATFORM, resolved by PRISMA_MIGRATION_ENGINE_BINARY)
Introspection Engine  : introspection-core ENGINE_VERSION (at sanitized_path/introspection-engine-TEST_PLATFORM, resolved by PRISMA_INTROSPECTION_ENGINE_BINARY)
Format Binary         : prisma-fmt ENGINE_VERSION (at sanitized_path/prisma-fmt-TEST_PLATFORM, resolved by PRISMA_FMT_BINARY)
Format Wasm           : @prisma/prisma-fmt-wasm CLI_VERSION.ENGINE_VERSION
Default Engines Hash  : ENGINE_VERSION
Studio                : STUDIO_VERSION
`;
